### DISCUSSION TITLE
Two server url

### DISCUSSION THREAD

**Discussion by alexlansovich** [aldas/go-modbus-client#13] [Q&A]
Hi.

Why in your example we need to write two the same server url? Also first sever can be any like 1.1.1.1 -and this is working. )
But without first url - dont work. Can you explain logic?
Also  why we duplicate unit ID in add request?

b := modbus.NewRequestBuilder(**"localhost:5020",** 1) - can be (**"1.1.1.1:5020"**, 1) 

requests, _ := b.Add(b.Uint16(18).UnitID(0).Name("test_do")).
    Add(b.Int64(18).Name("alarm_do_1").UnitID(0)). **(why we dupilcate UnitID? )**
    ReadHoldingRegistersTCP() // split added fields into multiple requests with suitable quantity size

client := modbus.NewTCPClient()
if err := client.Connect(context.Background(), **"localhost:5020"**); err != nil {
    return err
}

**Comment by aldas** [comment]
When you create [Builder](https://github.com/aldas/go-modbus-client/blob/9804df35caf7d6caf51b0131303919cfb1d918b1/builder.go#L162) instance with `modbus.NewRequestBuilder("localhost:5020", 1) ` that first argument is used as default [address](https://github.com/aldas/go-modbus-client/blob/9804df35caf7d6caf51b0131303919cfb1d918b1/builder.go#L48) for requests that builder creates.  You can create multiple different request to multiple different servers with same builder.

For example this would create 2 [requests](https://github.com/aldas/go-modbus-client/blob/9804df35caf7d6caf51b0131303919cfb1d918b1/builder.go#L366) object to 2 different servers for same register
```go
requests, _ := b.Add(b.Uint16(18).UnitID(0).Name("test_do").ServerAddress("1.1.1.1:502")).
    Add(b.Uint16(18).UnitID(0).Name("test_do2").ServerAddress("1.1.1.2:502")).
    ReadHoldingRegistersTCP() 
```

same goes for `Unit()` as single modbus request can only be sent to single unit/slave instance, but you could have multiple registers for different units/slaves instance that you want to request. Builder intention is to help you to split your register into minimal amount of request that need to be send to modbus server(s).


in that example that you are talking we only connect to single server. This is for simplicity sake:
```go
if err := client.Connect(context.Background(), "localhost:5020"); err != nil {
    return err
}
```

**Reply by aldas** [reply]
I see that you starred my PHP library. Basically Go version has the same idea for builder - it helps you to split registers into requests  across different server or unitID/slaves or when register amount is bigger than single modbus request can hold.

For PHP there is couple examples that create connection for each split request. See https://github.com/aldas/modbus-tcp-client/blob/master/examples/example_parallel_requests_reactphp.php    in Go version you can do the same but my example only connects to single and assumes that all requests that the `for` loops over are for the same server

**Comment by alexlansovich** [comment]
OK. if i need only one server.
How i can make this with one url? without duplicate server.

```go
package main

import (
	"context"
	"fmt"
	"github.com/aldas/go-modbus-client"
)

func main() {

	serverAddress := "172.25.28.21:8899" // Виносимо адресу в змінну
	b := modbus.NewRequestBuilder(serverAddress, 1)

	requests, _ := b.Add(b.Uint16(10240).Name("main_volt_a")).
		Add(b.Uint16(10242).Name("main_volt_b")).
		Add(b.Uint16(10244).Name("main_volt_c")).
		ReadHoldingRegistersTCP()
	client := modbus.NewTCPClient()
	if err := client.Connect(context.Background(),serverAddress); err != nil {
		fmt.Printf("error backgroud\n")
		return
	}
	for _, req := range requests {
		resp, err := client.Do(context.Background(), req)
		if err != nil {
			fmt.Printf("error request\n")
			return
		}

		// extract values to FieldValue struct
		fields, _ := req.ExtractFields(resp.(modbus.RegistersResponse), true)
		for _, t := range fields {
			fmt.Print(t.Field.Name, "=", t.Value,"\n")
		}
	}
}
```

**Reply by aldas** [reply]
As you have only single client instance and that one is connected to `"172.25.28.21:8899"` this example should be enough.

This is my output for that same code (only address is `127.0.0.1:5020`, testing against [diagslave](https://www.modbusdriver.com/diagslave.html) server.

```bash
x@x:~/code$ ./diagslave -m tcp -p 5020
diagslave 3.1 - FieldTalk(tm) Modbus(R) Diagnostic Slave Simulator
Copyright (c) 2002-2018 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP
Slave configuration: address = -1, master activity t/o = 3.00
IP configuration: port = 5020, connection t/o = 60.00

Server started up successfully.
Listening to network (Ctrl-C to stop)
.
validateMasterIpAddr: accepting connection from 127.0.0.1
Slave   1: readHoldingRegisters from 10241, 5 references
....^C
```

**Comment by alexlansovich** [comment]
Yes. my example is working.
I thinking about code optimization and structure.

In global - i will get all devices with all them registers.
every device will run own go routine with array of registers.
go routine will create run a loop with 10 seconds sleep (for example) with requests.
So i need to create with Add all registers.

```
-> one of go routine start
func main() {

	serverAddress := "172.25.28.21:8899"
	b := modbus.NewRequestBuilder(serverAddress, 1)

	requests, _ := b.Add(b.Uint16(10240).Name("main_volt_a")).
		Add(b.Uint16(10242).Name("main_volt_b")).
		Add(b.Uint16(10244).Name("main_volt_c")).
		ReadHoldingRegistersTCP()
	client := modbus.NewTCPClient()

-> main loop start
	if err := client.Connect(context.Background(),serverAddress); err != nil {
		fmt.Printf("error backgroud\n")
		return
	}
	for _, req := range requests {
		resp, err := client.Do(context.Background(), req)
		if err != nil {
			fmt.Printf("error request\n")
			return
		}
		fields, _ := req.ExtractFields(resp.(modbus.RegistersResponse), true)
		for _, t := range fields {
			fmt.Print(t.Field.Name, "=", t.Value,"\n")
                 -> add values to mqtt for example.
		}
	}
-> main loop end.

}

```
If my device will gone offline(and then online after some time) - will i need to make new client.Connect(context.Background(),serverAddress)?
